packages/wrangler/src/dev/start-server.ts

```diff
@@ -131,26 +143,62 @@ export async function startDevServer(
 
 				return { accountId, apiToken: requireApiToken() };
 			},
+			persist: props.localPersistencePath ?? undefined,
+			testScheduled: props.testScheduled,
+			registry: workerDefinitions,
 		},
 		build: {
-			// format: props.entry.format,
+			bundle: !props.noBundle,
+			define: props.define,
+			jsxFactory: props.jsxFactory,
+			jsxFragment: props.jsxFragment,
+			tsconfig: props.tsconfig,
+			minify: props.minify,
+			processEntrypoint: props.processEntrypoint,
+			additionalModules: props.additionalModules,
 			moduleRoot: props.entry.moduleRoot,
-			nodejsCompatMode: null,
+			moduleRules: props.rules,
+			nodejsCompatMode: props.nodejsCompatMode,
 		},
 	};
 
-	// temp: fake these events by calling the handler directly
-	if (!props.experimentalDevEnv) {
-		devEnv.proxy.onConfigUpdate({
-			type: "configUpdate",
-			config: fakeResolvedInput(startDevWorkerOptions),
-		});
-		devEnv.proxy.onBundleStart({
-			type: "bundleStart",
-			config: fakeResolvedInput(startDevWorkerOptions),
+	if (props.experimentalDevEnv) {
+		// to comply with the current contract of this function, call props.onReady on reloadComplete
+		devEnv.runtimes.forEach((runtime) => {
+			runtime.on("reloadComplete", async (ev) => {
+				const { proxyWorker } = await devEnv.proxy.ready.promise;
+				const url = await proxyWorker.ready;
+
+				props.onReady?.(url.hostname, parseInt(url.port), ev.proxyData);
+			});
 		});
+
+		await devEnv.config.set(startDevWorkerOptions);
+
+		await Promise.race([
+			once(devEnv.runtimes[0], "reloadComplete"),
+			once(devEnv.runtimes[1], "reloadComplete"),
+			once(devEnv.runtimes[0], "error").then((err) => Promise.reject(err)),
+			once(devEnv.runtimes[1], "error").then((err) => Promise.reject(err)),
+		]);
+
+		return {
+			stop: async () => {
+				await Promise.allSettled([stopWorkerRegistry(), devEnv.teardown()]);
+			},
+		};
```

Comment:
Made this function clearer by returning early when props.experimentalDevEnv is set. The rest of the function is for when it is not set